Docs: Tutorial for streaming ingestion using Kafka + Docker file to use with Jupyter tutorials (#13984)
apache/druid: docs/tutorials/tutorial-jupyter-docker.md

Docker Compose references the configuration in `docker-compose.yaml`.
Running Druid in Docker also requires the `environment` file, which
sets the configuration properties for the Druid services.
To get started, download both `docker-compose.yaml` and `environment` from
[`tutorial-jupyter-docker.zip`](https://github.com/apache/druid/blob/master/examples/quickstart/jupyter-notebooks/docker-jupyter/tutorial-jupyter-docker.zip).

Alternatively, you can clone the [Apache Druid repo](https://github.com/apache/druid) and
access the files in `druid/examples/quickstart/jupyter-notebooks/docker-jupyter`.

### Start only the Jupyter container

If you already have Druid running locally, you can run only the Jupyter container to complete the tutorials.
In the same directory as `docker-compose.yaml`, start the application:

```bash
docker-compose --profile jupyter up -d
```

The port assigned to Jupyter is `8889` by default.
You can override the port number by setting the `JUPYTER_PORT` environment variable before starting the Docker application.

### Start Jupyter and Druid

Running Druid in Docker requires the `environment` file as well as the `DRUID_VERSION` environment variable.

> `DRUID_VERSION` supplies the Docker tag that specifies which version of Druid to pull from the

Reviewer: ```suggestion
`DRUID_VERSION` references the Docker tag for the version of Druid to pull from the
```

[Apache Druid Docker Hub](https://hub.docker.com/r/apache/druid/tags).

In the same directory as `docker-compose.yaml` and `environment`, start the application:

```bash
DRUID_VERSION=25.0.0 docker-compose --profile druid-jupyter up -d
```

### Start Jupyter, Druid, and Kafka

Running Druid in Docker requires the `environment` file as well as the `DRUID_VERSION` environment variable.

In the same directory as `docker-compose.yaml` and `environment`, start the application:

```bash
DRUID_VERSION=25.0.0 docker-compose --profile all-services up -d
```

### Use locally built image

The default Docker Compose file pulls the custom Jupyter Notebook image from Imply's Docker Hub.

> If you would rather build the image locally, do the following:

Reviewer: ```suggestion
If you prefer to build the image locally, do the following:
```
avoid conditional

1. Clone the Apache Druid repository.
2. Navigate to `examples/quickstart/jupyter-notebooks/docker-jupyter`.
3. Start the services using `-f docker-compose-local.yaml` in the `docker-compose` command. For example:

```bash
DRUID_VERSION=25.0.0 docker-compose --profile all-services -f docker-compose-local.yaml up -d
```

## Access Jupyter-based tutorials

At startup, Docker creates and mounts a volume to persist data from the container to your local machine.
Access the files created in the Docker container in the `notebooks` folder in your local working directory.

1. Navigate to available notebooks at http://localhost:8889.
   > If you set `JUPYTER_PORT` to another port number, replace `8889` with the value of the Jupyter port.

2. Select a tutorial.

> 3. From the navigation menu, select **File > Save as...**.

Reviewer: you only need to do this if you want to edit the notebook? If so, Maybe make an optional step?

In the **Save As** dialog, enter `work/<notebook name>.ipynb`. This step allows you to retain a local copy of your work in the notebook. If the notebook still displays as read only, you may need to refresh the page in your browser.

## View the Druid web console

To access the Druid web console in Docker, go to http://localhost:8888/unified-console.html.
Use the web console to view datasources and ingestion tasks that you create in the tutorials.

## Stop Docker containers

Shut down the Docker application using the following command:

```bash
docker-compose down -v
```

## Tutorial setup without using Docker

To use the Jupyter Notebook-based tutorials without using Docker, do the following:

1. Clone the Apache Druid repo, or download the [tutorials](tutorial-jupyter-index.md#tutorials)
as well as the [Python client for Druid](tutorial-jupyter-index.md#python-api-for-druid).

> 2. Install the prerequisite Python packages with the following commands:

Reviewer: Add note that individual notebooks may require you to install additional packages and that they'll be listed w/in the notebook

  ~ regarding copyright ownership.  The ASF licenses this file
  ~ to you under the Apache License, Version 2.0 (the
  ~ "License"); you may not use this file except in compliance
  ~ with the License.  You may obtain a copy of the License at
  ~
  ~   http://www.apache.org/licenses/LICENSE-2.0
  ~
  ~ Unless required by applicable law or agreed to in writing,
  ~ software distributed under the License is distributed on an
  ~ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  ~ KIND, either express or implied.  See the License for the
  ~ specific language governing permissions and limitations
  ~ under the License.
  -->

Apache Druid provides a custom Jupyter container that contains the prerequisites
for all Jupyter-based Druid tutorials, as well as all of the tutorials themselves.
You can run the Jupyter container, as well as containers for Druid and Apache Kafka,
using the Docker Compose file provided in the Druid GitHub repository.

You can run the following combination of applications:
* [Jupyter only](#start-only-the-jupyter-container)
* [Jupyter and Druid](#start-jupyter-and-druid)
* [Jupyter, Druid, and Kafka](#start-jupyter-druid-and-kafka)

## Prerequisites

> Jupyter in Docker requires that you have **Docker** and **Docker Compose**.

Reviewer: **Docker Compose** is included in current versions of desktop:
https://docs.docker.com/get-started/08_using_compose/#install-docker-compose

Author: I think people could install these separately if they wanted but I included that the suggested method is Docker Desktop:
> We recommend installing these through [Docker Desktop](https://docs.docker.com/desktop/).

## Launch the Docker containers

You run Docker Compose to launch Jupyter and optionally Druid or Kafka.
Docker Compose references the configuration in `docker-compose.yaml`.
Running Druid in Docker also requires the `environment` file, which
sets the configuration properties for the Druid services.
To get started, download both `docker-compose.yaml` and `environment` from
[`tutorial-jupyter-docker.zip`](https://github.com/apache/druid/blob/master/examples/quickstart/jupyter-notebooks/docker-jupyter/tutorial-jupyter-docker.zip).

Alternatively, you can clone the [Apache Druid repo](https://github.com/apache/druid) and
access the files in `druid/examples/quickstart/jupyter-notebooks/docker-jupyter`.

### Start only the Jupyter container

If you already have Druid running locally, you can run only the Jupyter container to complete the tutorials.
In the same directory as `docker-compose.yaml`, start the application:

```bash
docker compose --profile jupyter up -d
```

The Docker Compose file assigns `8889` for the Jupyter port.
You can override the port number by setting the `JUPYTER_PORT` environment variable before starting the Docker application.

### Start Jupyter and Druid

> Running Druid in Docker requires the `environment` file as well as the `DRUID_VERSION` environment variable.

Reviewer 2: the --> an?

We've not talked about an environment file yet so I don't know which one is "the" environment file.

`DRUID_VERSION` references the Docker tag for the version of Druid to pull from the
[Apache Druid Docker Hub](https://hub.docker.com/r/apache/druid/tags).

In the same directory as `docker-compose.yaml` and `environment`, start the application:

```bash
DRUID_VERSION={{DRUIDVERSION}} docker compose --profile druid-jupyter up -d
```

### Start Jupyter, Druid, and Kafka

Running Druid in Docker requires the `environment` file as well as the `DRUID_VERSION` environment variable.

In the same directory as `docker-compose.yaml` and `environment`, start the application:

```bash
DRUID_VERSION={{DRUIDVERSION}} docker compose --profile all-services up -d
```

### Update image from Docker Hub

> If you already have a local cache of the Jupyter image, you can update the image before running the application using the following command:

Reviewer 2: When might I want to update the image? Doesn't Docker grab the latest one itself?

Author: From what I can tell, you can set up tooling to automatically update to the latest image, but this doesn't happen by default... https://stackoverflow.com/questions/26423515/how-to-automatically-update-your-docker-containers-if-base-images-are-updated
